# 2. teden: Numpy - Domača naloga

Uvozimo modul numpy:

In [2]:
import numpy as np

### 1.Create a null vector of size 10

In [2]:
np.zeros(10)

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

Odg: `[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]`

### 2.Create a null vector of size 10 but the fifth value which is 1

In [6]:
v = np.zeros(10)
v[4] = 1
print(v)

[0. 0. 0. 0. 1. 0. 0. 0. 0. 0.]


Odg: `[0. 0. 0. 0. 1. 0. 0. 0. 0. 0.]`

### 3.Create a vector with values ranging from 10 to 49

In [10]:
np.arange(10,50)

array([10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26,
       27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43,
       44, 45, 46, 47, 48, 49])

Odg: `[10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49]`

### 4.Reverse a vector (first element becomes last)

In [12]:
Z = np.arange(50)

In [13]:
np.flip(Z)

array([49, 48, 47, 46, 45, 44, 43, 42, 41, 40, 39, 38, 37, 36, 35, 34, 33,
       32, 31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16,
       15, 14, 13, 12, 11, 10,  9,  8,  7,  6,  5,  4,  3,  2,  1,  0])

Odg: `[49 48 47 46 45 44 43 42 41 40 39 38 37 36 35 34 33 32 31 30 29 28 27 26 25 24 23 22 21 20 19 18 17 16 15 14 13 12 11 10  9  8  7  6  5  4  3  2 1  0]`

### 5.Create a 3x3 matrix with values ranging from 0 to 8

**Pomoč**: metoda .reshape() -> [Dokumentacija](https://docs.scipy.org/doc/numpy/reference/generated/numpy.reshape.html)

In [14]:
np.arange(9).reshape((3,3))

array([[0, 1, 2],
       [3, 4, 5],
       [6, 7, 8]])

Odg: 

`[[0 1 2]
 [3 4 5]
 [6 7 8]]`

### 6.Find indices of non-zero elements from [1,2,0,0,4,0]

**Pomoč:** Google, Stack Overflow, Numpy dokumentacija

In [16]:
a = [1,2,0,0,4,0]

In [17]:
np.nonzero(a)

(array([0, 1, 4]),)

Odg: `(array([0, 1, 4], dtype=int64),)`

### 7.Create a 3x3 identity matrix

In [18]:
np.eye(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

Odg: 

`[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]`

### 8.Create a 3x3x3 array with random values

In [20]:
np.random.random((3,3,3))

array([[[0.61358539, 0.20156856, 0.89768976],
        [0.1789916 , 0.18892879, 0.31935495],
        [0.39831434, 0.95449854, 0.92955086]],

       [[0.16510227, 0.1541347 , 0.88787236],
        [0.38244363, 0.40238998, 0.1976831 ],
        [0.8032543 , 0.90610625, 0.51742132]],

       [[0.42553279, 0.1959134 , 0.8359994 ],
        [0.13401708, 0.72275069, 0.56676428],
        [0.48919493, 0.58922398, 0.84593096]]])

Odgovor npr.: 
    
`[[[0.61712056 0.38943629 0.16277319]
  [0.19481511 0.12725415 0.6717974 ]
  [0.15340477 0.78713636 0.52864134]]

 [[0.91153827 0.7756351  0.88600077]
  [0.03509814 0.79676837 0.39167056]
  [0.55410251 0.67583417 0.64169036]]

 [[0.46862042 0.05954143 0.42512482]
  [0.34746791 0.97653267 0.0024134 ]
  [0.53522577 0.37192358 0.29601563]]]`

### 9.Create a 10x10 array with random values and find the minimum and maximum values

In [33]:
A = np.random.random((10,10))



In [36]:
Zmin = np.min(A)
Zmax = np.max(A)
print(Zmin)
print(Zmax)

0.018026053530285746
0.9782703047773241


Odg: `Zmin= 0.004874889654358006 Zmax= 0.992254173881756`

### 10.Create a 2d array (10x10) with 1 on the border and 0 inside

In [37]:
x = np.ones((10,10))

In [38]:
x[1:-1,1:-1] = 0
print(x)


[[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]]


Odg: 
    
`[[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]]`

### 11.What is the result of the following expression?

In [39]:
0 * np.nan

nan

In [40]:
np.nan == np.nan

False

In [41]:
np.inf > np.nan

False

In [42]:
np.nan - np.nan

nan

In [43]:
0.3 == 3 * 0.1 

False

### 12.Create a 8x8 matrix and fill it with a checkerboard pattern

In [58]:
x = np.zeros((8,8))
x[1::2,::2] = 1
x[::2,1::2] = 1
print(x)


[[0. 1. 0. 1. 0. 1. 0. 1.]
 [1. 0. 1. 0. 1. 0. 1. 0.]
 [0. 1. 0. 1. 0. 1. 0. 1.]
 [1. 0. 1. 0. 1. 0. 1. 0.]
 [0. 1. 0. 1. 0. 1. 0. 1.]
 [1. 0. 1. 0. 1. 0. 1. 0.]
 [0. 1. 0. 1. 0. 1. 0. 1.]
 [1. 0. 1. 0. 1. 0. 1. 0.]]


In [55]:
x

array([0., 1., 0., 1., 0., 1., 0., 1., 0., 1., 0., 1., 0., 1., 0., 1., 0.,
       1., 0., 1., 0., 1., 0., 1., 0., 1., 0., 1., 0., 1., 0., 1., 0., 1.,
       0., 1., 0., 1., 0., 1., 0., 1., 0., 1., 0., 1., 0., 1., 0., 1., 0.,
       1., 0., 1., 0., 1., 0., 1., 0., 1., 0., 1., 0., 1.])

Odg: 

`[[0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]]`

### 13.Given a 1D array, negate all elements which are between 3 and 8, in place.

In [59]:
Z = np.arange(11)

In [60]:
Z

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

In [63]:
m = (Z >3) & (Z < 8)


In [64]:
Z[m] *= -1
print(Z)

[ 0  1  2  3 -4 -5 -6 -7  8  9 10]


Odg: `[ 0  1  2  3 -4 -5 -6 -7 -8  9 10]`

### 14.Create a vector of size 10 with values ranging from 0 to 1, both excluded

In [76]:
x = np.linspace(0,1,12,endpoint=True)[1:-1]


In [77]:
x

array([0.09090909, 0.18181818, 0.27272727, 0.36363636, 0.45454545,
       0.54545455, 0.63636364, 0.72727273, 0.81818182, 0.90909091])

Odg: `[0.09090909 0.18181818 0.27272727 0.36363636 0.45454545 0.54545455 0.63636364 0.72727273 0.81818182 0.90909091]`

### 15.Create a random vector of size 10 and sort it

In [85]:
x = np.random.random(10)


In [86]:
np.sort(x)

array([0.24320978, 0.25878678, 0.35201833, 0.40743826, 0.52357198,
       0.56499541, 0.57556966, 0.60418526, 0.63485527, 0.75310394])

Odg. npr.: `[0.08755811 0.17786018 0.45041482 0.48746541 0.68379823 0.70209576 0.77273884 0.86106446 0.93083351 0.98349603]`

### 16. Consider two random array A anb B, check if they are equal

In [3]:
A = np.random.randint(0,2,5)

In [4]:
B = np.random.randint(0,2,5)

In [6]:
np.array_equal(A, B)

False

### 17.Make an array immutable (read-only)

In [7]:
Z = np.zeros(10)

In [8]:
Z.flags.writeable = False

In [9]:
Z[0] = 1

ValueError: assignment destination is read-only

### 18.Create random vector of size 10 and replace the maximum value by 0

In [26]:
A = np.random.randint(1,10,10)

In [27]:
Amax = np.max(A)
A[A == Amax] = 0
print(A)
print(Amax)


[8 3 6 8 0 6 3 2 4 4]
9


### 19.How to find the closest value (to a given scalar -v) in an array -Z?

In [35]:
Z = np.arange(100)

In [36]:
v = np.random.uniform(0,100)

In [37]:
print(Z)
print(v)
index = (np.abs(Z - v)).argmin()

print(Z[index])

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47
 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71
 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95
 96 97 98 99]
40.100856309757205
40


### 20.How to convert an integer (32 bits)  array into a float (32 bits) in place?

In [38]:
Z = np.arange(10, dtype=np.int32)

In [39]:
print(Z)

[0 1 2 3 4 5 6 7 8 9]


In [41]:
np.float32(Z)

array([0., 1., 2., 3., 4., 5., 6., 7., 8., 9.], dtype=float32)

Odg: `[0. 1. 2. 3. 4. 5. 6. 7. 8. 9.]`

### 21.How to find the most frequent value in an array?

**Pomoč:** metoda .bincount() -> [Dokumentacija](https://docs.scipy.org/doc/numpy/reference/generated/numpy.bincount.html?highlight=bincount#numpy.bincount)

In [46]:
Z = np.random.randint(0,10,50)
print(Z)

[6 3 4 9 4 4 8 2 0 9 7 8 5 7 7 5 0 1 5 8 9 7 5 4 4 9 5 4 3 6 8 1 3 5 2 8 3
 6 1 3 9 8 3 3 5 1 4 2 9 4]


In [48]:

np.bincount(Z).argmax()

4